authentik/policies/geoip/models.py

```diff
@@ -0,0 +1,81 @@
+"""GeoIP policy"""
+
+from django.contrib.postgres.fields import ArrayField
+from django.db import models
+from django.utils.translation import gettext as _
+from rest_framework.serializers import BaseSerializer
+
+from authentik.policies.exceptions import PolicyException
+from authentik.policies.geoip.countries import COUNTRIES
+from authentik.policies.geoip.exceptions import GeoIPNotFoundException
+from authentik.policies.models import Policy
+from authentik.policies.types import PolicyRequest, PolicyResult
+
+
+class GeoIPPolicyMode(models.TextChoices):
+    """Block or only allow connections from entities"""
+
+    ALLOW = "allow", _("Only allow connections from entities")
+    BLOCK = "block", _("Block connections from entities")
+
+
+class GeoIPPolicy(Policy):
+    """Ensure the user satisfies requirements of geography or network topology, based on IP
+    address."""
+
+    asn_mode = models.TextField(choices=GeoIPPolicyMode.choices)
+    country_mode = models.TextField(choices=GeoIPPolicyMode.choices)
+    asn_list = ArrayField(models.IntegerField(), blank=True, default=list)
+    country_list = ArrayField(
+        models.CharField(choices=COUNTRIES, max_length=2), blank=True, default=list
+    )
+
+    @property
+    def serializer(self) -> type[BaseSerializer]:
+        from authentik.policies.geoip.api import GeoIPPolicySerializer
+
+        return GeoIPPolicySerializer
+
+    @property
+    def component(self) -> str:  # pragma: no cover
+        return "ak-policy-geoip-form"
+
+    def passes(self, request: PolicyRequest) -> PolicyResult:
+        """TODO"""
+
+        # This is not a single get chain because `request.context` can contain `{ "asn": None }`.
+        asn_data = request.context.get("asn")
+        geoip_data = request.context.get("geoip")
+        asn = asn_data.get("asn") if asn_data else None
+        country = geoip_data.get("country") if geoip_data else None
+
+        if not asn:
```

Comment:
both of these cases can also happen if ASN/GeoIP is not configured (pretty rare since we include the DB by default, but it is possible to disable)

Reply:
(We've had a bit of an offline discussion, but no resolution to this.) I kept the behavior for now, this way we don't leak information on whether a GeoIP database is available.

Let me know if you don't agree.